### ISSUE TITLE
Duplicate Vendor/Customer Order Emails and Order Notes

### ISSUE THREAD

**Issue by dwh10045** [woocommerce/woocommerce-gateway-stripe#3501]
I am currently using WooCommerce Stripe Gateway plugin version 8.7.0
I also am using WooCommerce plugin version 9.3.3
I am also using the WC Vendors Stripe Connect plugin version 2.2.2.4
I am using WordPress version 6.5.5

I am in the final stages of development of my website, so this includes testing the payment processing between vendors and buyers, as well as checking email notifications.

Whenever I process a payment, I am receiving duplicate Order emails to both the Vendor and the Buyers, as well as multiple Order notes. 

Here are some screenshots of the email messages:

![Image](https://github.com/user-attachments/assets/21833a85-e501-4caf-8a94-fd64c487616c)

Here is a screenshot of the Order Notes:

![Image](https://github.com/user-attachments/assets/ba921f60-20b0-4cf5-8512-af8670baac09)

Here is a screenshot of the logs from Stripe:

![Image](https://github.com/user-attachments/assets/fdeca52f-b7d6-4a84-92bf-61f3520d587e)

Please advise.

**Comment by paulkirspuu** [comment]
Try WooCommerce Stripe Gateway plugin version **8.6.0**.

8.7.0 looks to be missing https://github.com/woocommerce/woocommerce-gateway-stripe/commit/4468d290d774aa61f8d8b5134e5ce70cbc3a6b25.

**Comment by chrism245** [comment]
8902186-zen

**Comment by KevinPlusPlus** [comment]
8902186-zen may be my ticket from Woo Commerce support, but I experienced this same issue after updating to 8.8 and downgrading back to 8.7 resolved it.

**Comment by solaceten** [comment]
Hello.  Adding to this ticket because I have the same issue with two different websites.  

On both sites, I am running the new Stripe setup v.8.8.0 (just changed over), I am not using Legacy setup.

I received two (duplicates) and in the order notes, I see duplicated entries for "Stripe charge complete (Charge ID" and also duplicates entries for "Order status changed from Pending payment to Completed."

Customer is only receiving single email, but admin gets two [new order] emails.

I am running WP in WordPress posts storage (legacy) mode, b ut with compatibility mode enabled - maybe that is the issue?  Two order tables essentially??

What is the fix? Should I downgrade back to 8.7?

Also - I see the issues ia also being address[ in this ticket](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2463#issuecomment-2117274970)

**Comment by KevinPlusPlus** [comment]
For what it's worth, I am using HPOS and not duplicating to the legacy tables.

**Comment by solaceten** [comment]
Hi @KevinPlusPlus - I did sync to HPOS but had a conflicting plugin that I can't remove just yet, hence running with compatibility mode .

However, in the second site, it's also running in compatibility mode (HPOS synced) - and virtually identical to site 1.  However, it is not getting the duplicate order notes or the duplicate order emails... very odd. 

Also, I am not running 'WC Vendors Stripe Connect' that the OP is running, so I don't believe that's relevant!

[Also linking in this thread ](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2331 ) (also prematurely closed?)

**Comment by solaceten** [comment]
@KevinPlusPlus are you using Legacy or upgraded checkout?

**Comment by solaceten** [comment]
@Mayisha @a-danae Are you able to assign this?

**Comment by KevinPlusPlus** [comment]
@solaceten I am using the blocks-based checkout and the most updated Stripe checkout experience.

**Comment by Mayisha** [comment]
Thanks for sharing all the details. We are looking into this and trying to reproduce it on our end.
Could you please confirm the payment method that was used in these orders?

**Comment by KevinPlusPlus** [comment]
Visa and Mastercard via Stripe, it was happening on every single order.

**Comment by solaceten** [comment]
@Mayisha in my case they are all stripe payments with Visa

Here is my WP status report with some redactions

WordPress Environment<br><br>WordPress address (URL): example.com<br>Site address (URL): example.com<br>WC Version: 9.3.3<br>Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.<br>Action Scheduler Version: ✔ 3.8.1<br>Log Directory Writable: ✔<br>WP Version: 6.6.2<br>WP Multisite: –<br>WP Memory Limit: 256 MB<br>WP Debug Mode: –<br>WP Cron: ✔<br>Language: en_US<br>External object cache: – Server Environment<br><br>Server Info: Apache<br>PHP Version: 8.1.30<br>PHP Post Max Size: 8 MB<br>PHP Time Limit: 90<br>PHP Max Input Vars: 1000<br>cURL Version: 7.61.1<br>OpenSSL/1.1.1k<br><br>SUHOSIN Installed: –<br>MySQL Version: 8.0.40<br>Max Upload Size: 8 MB<br>Default Timezone is UTC: ✔<br>fsockopen/cURL: ✔<br>SoapClient: ✔<br>DOMDocument: ✔<br>GZip: ✔<br>Multibyte String: ✔<br>Remote Post: ✔<br>Remote Get: ✔ Database<br><br>WC Database Version: 9.3.3<br>WC Database Prefix: wp_<br>Total Database Size: 182.17MB<br>Database Data Size: 134.88MB<br>Database Index Size: 47.29MB<br>wp_woocommerce_sessions: Data: 0.27MB + Index: 0.02MB + Engine MyISAM<br>wp_woocommerce_api_keys: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_attribute_taxonomies: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_downloadable_product_permissions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_order_items: Data: 0.32MB + Index: 0.18MB + Engine MyISAM<br>wp_woocommerce_order_itemmeta: Data: 4.32MB + Index: 2.10MB + Engine MyISAM<br>wp_woocommerce_tax_rates: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_tax_rate_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_shipping_zones: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_shipping_zone_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_shipping_zone_methods: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_payment_tokens: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_woocommerce_payment_tokenmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_woocommerce_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_actionscheduler_actions: Data: 1.25MB + Index: 0.46MB + Engine MyISAM<br>wp_actionscheduler_claims: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_actionscheduler_groups: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_actionscheduler_logs: Data: 0.73MB + Index: 0.53MB + Engine MyISAM<br>wp_aryo_activity_log: Data: 8.93MB + Index: 4.38MB + Engine MyISAM<br>wp_blc_filters: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_blc_instances: Data: 0.02MB + Index: 0.01MB + Engine MyISAM<br>wp_blc_links: Data: 0.22MB + Index: 0.04MB + Engine MyISAM<br>wp_blc_synch: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_commentmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_comments: Data: 8.43MB + Index: 3.18MB + Engine MyISAM<br>wp_gla_attribute_mapping_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB<br>wp_gla_budget_recommendations: Data: 0.19MB + Index: 0.11MB + Engine InnoDB<br>wp_gla_merchant_issues: Data: 0.02MB + Index: 0.00MB + Engine InnoDB<br>wp_gla_shipping_rates: Data: 0.02MB + Index: 0.03MB + Engine InnoDB<br>wp_gla_shipping_times: Data: 0.02MB + Index: 0.02MB + Engine InnoDB<br>wp_hurrytimer_evergreen: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_iwp_backup_status: Data: 0.10MB + Index: 0.00MB + Engine MyISAM<br>wp_iwp_file_list: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_iwp_processed_iterator: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_kbp_countdown_entry: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_kbp_form_entry: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_kbp_form_entrymeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_kbp_form_events: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_kbp_query_index: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_ksp_layers: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_ksp_sliders: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_ksp_slides: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_links: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_options: Data: 7.28MB + Index: 0.49MB + Engine MyISAM<br>wp_pmxe_exports: Data: 0.37MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxe_google_cats: Data: 0.26MB + Index: 0.05MB + Engine MyISAM<br>wp_pmxe_posts: Data: 0.25MB + Index: 0.11MB + Engine MyISAM<br>wp_pmxe_templates: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxi_files: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxi_hash: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxi_history: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxi_images: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxi_imports: Data: 0.07MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxi_posts: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_pmxi_templates: Data: 0.02MB + Index: 0.00MB + Engine MyISAM<br>wp_postmeta: Data: 27.52MB + Index: 11.22MB + Engine MyISAM<br>wp_posts: Data: 16.15MB + Index: 1.21MB + Engine MyISAM<br>wp_smush_dir_images: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_term_relationships: Data: 0.01MB + Index: 0.03MB + Engine MyISAM<br>wp_term_taxonomy: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_termmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_terms: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_tm_taskmeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_tm_tasks: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_usermeta: Data: 0.27MB + Index: 0.15MB + Engine MyISAM<br>wp_users: Data: 0.01MB + Index: 0.02MB + Engine MyISAM<br>wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.01MB + Engine MyISAM<br>wp_wc_admin_notes: Data: 0.09MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_category_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_customer_lookup: Data: 0.11MB + Index: 0.08MB + Engine MyISAM<br>wp_wc_download_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_order_addresses: Data: 0.66MB + Index: 0.76MB + Engine MyISAM<br>wp_wc_order_coupon_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_order_operational_data: Data: 0.73MB + Index: 0.25MB + Engine MyISAM<br>wp_wc_order_product_lookup: Data: 0.46MB + Index: 0.36MB + Engine MyISAM<br>wp_wc_order_stats: Data: 0.35MB + Index: 0.21MB + Engine MyISAM<br>wp_wc_order_tax_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_orders: Data: 2.52MB + Index: 1.14MB + Engine InnoDB<br>wp_wc_orders_meta: Data: 15.55MB + Index: 16.11MB + Engine InnoDB<br>wp_wc_product_attributes_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_product_download_directories: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_wc_product_meta_lookup: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_wc_rate_limits: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_reserved_stock: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wc_tax_rate_classes: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_wc_webhooks: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wcpdf_invoice_number: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wcpdf_packing_slip_number: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wpfb_post_templates: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_wpfb_reviews: Data: 0.10MB + Index: 0.01MB + Engine MyISAM<br>wp_wpfb_total_averages: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_wpml_mails: Data: 26.52MB + Index: 0.06MB + Engine InnoDB<br>wp_wpmm_subscribers: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_yoast_indexable: Data: 0.04MB + Index: 0.02MB + Engine MyISAM<br>wp_yoast_indexable_hierarchy: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_yoast_migrations: Data: 0.00MB + Index: 0.01MB + Engine MyISAM<br>wp_yoast_primary_term: Data: 0.00MB + Index: 0.00MB + Engine MyISAM<br>wp_yoast_seo_links: Data: 0.04MB + Index: 0.01MB + Engine MyISAM Post Type Counts<br><br>aam_policy: 1<br>attachment: 173<br>aw_workflow: 3<br>custom_css: 1<br>flamingo_contact: 528<br>flamingo_inbound: 449<br>hurrytimer_countdown: 1<br>iwp_log: 4266<br>kadence_element: 1<br>kadence_wootemplate: 3<br>nav_menu_item: 40<br>oembed_cache: 1<br>page: 18<br>post: 25<br>product: 10<br>product_variation: 2<br>revision: 799<br>shop_coupon: 10<br>shop_order: 4768<br>shop_order_refund: 72<br>wp_navigation: 1<br>wpcf7_contact_form: 1<br><br>Secure connection (HTTPS): ✔<br>Hide errors from visitors: ✔ Active Plugins (39)<br><br>InfiniteWP - Client: by Revmakx – 1.13.0<br>Advanced Access Manager: by AAM – 6.9.42<br>Activity Log: by Activity Log Team – 2.11.0<br>Autoptimize: by Frank Goossens (futtta) – 3.1.12<br>Broken Link Checker: by WPMU DEV – 2.4.1<br>Checkout Countdown for WooCommerce: by Puri.io – 4.0.0<br>Contact Form 7: by Takayuki Miyoshi – 5.9.8<br>Flamingo: by Takayuki Miyoshi – 2.5<br>Google for WooCommerce: by WooCommerce – 2.8.6<br>Kadence Blocks - PRO Extension: by Kadence WP – 2.5.0<br>Kadence Blocks – Gutenberg Blocks for Page Builder Features: by Kadence WP – 3.3.2<br>Kadence Pro - Premium addon for the Kadence Theme: by Kadence WP – 1.0.21<br>Kadence Captcha: by Kadence WP – 1.3.2<br>Kadence WooCommerce Email Designer: by Kadence WP – 1.5.13<br>Limit Login Attempts Reloaded: by Limit Login Attempts Reloaded – 2.26.13<br>PW WooCommerce Copy Coupon: by Pimwick<br>Side Cart WooCommerce: by XootiX – 2.5.5<br>Hide SEO Bloat: by SO WP – 4.0.2<br>Woo Stickers by Webline: by Weblineindia – 1.2.3<br>PDF Invoices & Packing Slips for WooCommerce: by WP Overnight – 3.8.8<br>Subscribe to Newsletter for WooCommerce: by Kestrel – 4.1.1<br>WooCommerce: by Automattic – 9.3.3<br>Yoast SEO: by Team Yoast – 23.5<br>WP All Export Pro: by Soflyy – 1.9.1<br>WP Google Review Slider: by LJ Apps – 15.5<br>WP Last Login: by Konstantin Obenland – 7<br>WP Mail Logging: by WP Mail Logging Team – 1.13.1<br>WP-Optimize - Clean, Compress, Cache: by David Anderson<br>Ruhani Rabin<br>Team Updraft – 3.6.0<br>Smush: by WPMU DEV – 3.16.6<br>WP All Export - User Export Add-On Pro: by Soflyy – 1.0.7<br>WP All Export - WooCommerce Export Add-On Pro: by Soflyy – 1.0.9 Inactive Plugins (5)<br>Kadence Shop Kit: by Kadence WP – 2.2.6<br>LightStart - Maintenance Mode, Coming Soon and Landing Page Builder: by Themeisle – 2.6.13<br><br>Legacy API Enabled: –<br>Force SSL: –<br>Currency: NZD ($)<br>Currency Position: left<br>Thousand Separator: ,<br>Decimal Separator: .<br>Number of Decimals: 2<br>external (external)<br>grouped (grouped)<br>simple (simple)<br>variable (variable)<br>Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)<br>exclude-from-search (exclude-from-search)<br>featured (featured)<br>outofstock (outofstock)<br>rated-1 (rated-1)<br>rated-2 (rated-2)<br>rated-3 (rated-3)<br>rated-4 (rated-4)<br>rated-5 (rated-5)<br><br>Connected to WooCommerce.com: –<br>Enforce Approved Product Download Directories: –<br>HPOS feature enabled: –<br>Order datastore: WC_Order_Data_Store_CPT<br>HPOS data sync enabled: ✔ Logging<br><br>Enabled: ✔<br>Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2<br>Retention period: 90 days<br>Level threshold: Warning<br>Log directory size: 950 KB WC Pages<br><br>Shop base: # 326 - /shop/<br>Cart: # 327 - /shop/cart/ - Contains the [woocommerce_cart] shortcode<br>Checkout: # 328 - /shop/checkout/ - Contains the [woocommerce_checkout] shortcode<br>My account: # 329 - /shop/my-account/<br>Terms and conditions: # 600 - /shop/terms-conditions/ Theme<br><br>Name: === Kadence Child Theme ===<br>Version: 1.32<br>Author URL:<br>Child Theme: ✔<br>Parent Theme Name: Kadence<br>Parent Theme Version: 1.2.9<br>Parent Theme Author URL: https://www.kadencewp.com/<br>WooCommerce Support: ✔ Templates<br><br>Overrides: – Admin<br><br>Enabled Features: activity-panels<br>analytics<br>product-block-editor<br>coupons<br>core-profiler<br>customize-store<br>customer-effort-score-tracks<br>import-products-task<br>experimental-fashion-sample-products<br>shipping-smart-defaults<br>shipping-setting-tour<br>homescreen<br>mobile-app-banner<br>navigation<br>onboarding<br>onboarding-tasks<br>pattern-toolkit-full-composability<br>product-custom-fields<br>remote-inbox-notifications<br>remote-free-extensions<br>payment-gateway-suggestions<br>printful<br>shipping-label-banner<br>subscriptions<br>store-alerts<br>transient-notices<br>woo-mobile-welcome<br>wc-pay-promotion<br>wc-pay-welcome-page<br>launch-your-store<br>Disabled Features: experimental-blocks<br>marketing<br>minified-js<br>product-pre-publish-modal<br>settings<br>async-product-editor-category-field<br>product-editor-template-system<br>blueprint<br>reactify-classic-payments-settings<br><br>Daily Cron: ✔ Next scheduled: 2024-10-24 03:14:49 +13:00<br>Options: ✔<br>Notes: 175<br>Onboarding: completed Action Scheduler<br><br>Complete: 4,348<br>Oldest: 2024-09-23 13:56:20 +1200<br>Newest: 2024-10-24 11:11:50 +1300<br><br>Failed: 1<br>Oldest: 2024-10-05 16:12:07 +1300<br>Newest: 2024-10-05 16:12:07 +1300<br><br>Pending: 206<br>Oldest: 2024-10-24 11:56:05 +1300<br>Newest: 2025-03-07 23:59:59 +1300 Newsletter Subscription<br><br><br>Generated at: 2024-10-24 11:42:35 +13:00<br>

**Comment by KevinPlusPlus** [comment]
Just reporting in the 8..8.1 behaves the same for me, duplicating order notes and emails. Reverting to 8.7.0 restores normal functionality. So, no change from 8.8.0 to 8.8.1.

On a hunch I tried disabling Express Checkout for Apple Pay because a lot of the changes between 8.7 and 8.8 seemed related to that, but no change. And I wouldn't think Express Checkout would have any effect on subscription renewal orders anyway. But, thought it was worth testing.

I have programming, testing, and debugging skills and am glad to make relatively safe changes as experiments though it is a large, active site so don't want to do anything that is known to be code-breaking. A lot of the code changes from 8.7 to 8.8 were the setting of constants so that make it a bit more tedious to look for substantive code changes and try and narrow down a cause.

**Comment by solaceten** [comment]
Here is the order notes duplicated 

![Image](https://github.com/user-attachments/assets/67cf5134-3f1c-4068-823c-5789ee0b076e)

**Comment by KevinPlusPlus** [comment]
@solaceten That is exactly what we are seeing as well.

**Comment by solaceten** [comment]
Hi @KevinPlusPlus  - Not to mention the duplicate New Order email notifications that go to admin.....   Are you getting those too?

**Comment by KevinPlusPlus** [comment]
@solaceten Yes, duplicate emails to admins and to the customer.

Only a single Stripe charge, though; the charge itself is NOT being duplicated. Only the notes and corresponding events on the store are duplicated.

**Comment by solaceten** [comment]
Hi all - I have requested the other ticket to re-open as it has a lot more history on it.

https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2463

Suggest we merge the two ?

**Comment by n9yty** [comment]
Almost certainly this is due to some processing bug that is firing hooks/actions multiple times. Our store does monthly synchronized subscription renewals and we have custom logic that depends on those hooks/actions. It caused a massive problem this month that we had to go through and clean up manually order by order becuase the hooks fired multiple times per order.

**Comment by mattallan** [comment]
Hi @KevinPlusPlus @solaceten and @n9yty, I appreciate the discussion being had here and on #2463 and sorry for this on-going issue that is impacting your stores!

It's our top priority to get this issue resolved ASAP but we're still having trouble replicating it on our own stores.

Our current working theories are:
1. The `payment_intent.succeeded` webhook sent from Stripe is being processed while we are still processing the checkout payment in a separate request. This is causing both our webhook handler and the checkout process to mark the pending orders as completed in parallel.
2. The order locking mechanism that should prevent the issue mentioned in 1, is not working as intended on some stores.

For context, to prevent incoming webhooks from processing orders that we're still processing through the checkout, we lock the order by setting a 5-minute transient and then delete the transient when we're done.

Refactoring our code to avoid using transients for order locking or handling incoming webhook requests asynchronously are potential fixes we're looking at. However, our priority is to first reproduce the issue ourselves to ensure we're squashing this bug for good.

To assist us further:

1. Do you know of any plugins or custom code on your store that could be expiring transients immediately or affecting our code using short-lived transients?
2. Do you have Stripe logging enabled? If so, would you be comfortable sending me a copy of the Stripe logs that includes requests for an order with duplicate notes and emails? I'm in the [WooCommerce Community Slack channel](https://woocommerce.com/community-slack/) as `@mattallan` if you want to DM me the logs or my email is `matt.allan@automattic.com`.

I'll keep digging into this problem in the meantime. Thanks!

**Comment by solaceten** [comment]
Thank you @mattallan for the update, I don't think I have any custom code affecting transients. I do use Autoptimize plugin but I don't think that touches transients and I do run WP-Optimize manually - but haven't done so for a few weeks, so I can't see how that would cause this.

I have the exact same issue on two separate sites that share many of the following plugins mentioned in this Woo Status Report - hope it can help you? Some info redacted for privacy and I'm not logging Stripe right now, but will enable it. 

WordPress Environment

WordPress address (URL): https://example.com
Site address (URL): https://example.com
WC Version: 9.3.3
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.8.1
Log Directory Writable: ✔
WP Version: 6.6.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: – Server Environment

Server Info: Apache
PHP Version: 8.1.30
PHP Post Max Size: 8 MB
PHP Time Limit: 90
PHP Max Input Vars: 1000
cURL Version: 7.61.1
OpenSSL/1.1.1k

SUHOSIN Installed: –
MySQL Version: 8.0.40
Max Upload Size: 8 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔ Database

WC Database Version: 9.3.3
WC Database Prefix: wp_
Total Database Size: 182.17MB
Database Data Size: 134.88MB
Database Index Size: 47.29MB
wp_woocommerce_sessions: Data: 0.27MB + Index: 0.02MB + Engine MyISAM
wp_woocommerce_api_keys: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_attribute_taxonomies: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_downloadable_product_permissions: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_order_items: Data: 0.32MB + Index: 0.18MB + Engine MyISAM
wp_woocommerce_order_itemmeta: Data: 4.32MB + Index: 2.10MB + Engine MyISAM
wp_woocommerce_tax_rates: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_tax_rate_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zones: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_locations: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_shipping_zone_methods: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokens: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_woocommerce_payment_tokenmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_woocommerce_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_actionscheduler_actions: Data: 1.25MB + Index: 0.46MB + Engine MyISAM
wp_actionscheduler_claims: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_actionscheduler_groups: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_actionscheduler_logs: Data: 0.73MB + Index: 0.53MB + Engine MyISAM
wp_aryo_activity_log: Data: 8.93MB + Index: 4.38MB + Engine MyISAM
wp_blc_filters: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_blc_instances: Data: 0.02MB + Index: 0.01MB + Engine MyISAM
wp_blc_links: Data: 0.22MB + Index: 0.04MB + Engine MyISAM
wp_blc_synch: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_commentmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_comments: Data: 8.43MB + Index: 3.18MB + Engine MyISAM
wp_gla_attribute_mapping_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gla_budget_recommendations: Data: 0.19MB + Index: 0.11MB + Engine InnoDB
wp_gla_merchant_issues: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gla_shipping_rates: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gla_shipping_times: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_hurrytimer_evergreen: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_iwp_backup_status: Data: 0.10MB + Index: 0.00MB + Engine MyISAM
wp_iwp_file_list: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_iwp_processed_iterator: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_kbp_countdown_entry: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_kbp_form_entry: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_kbp_form_entrymeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_kbp_form_events: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_kbp_query_index: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_ksp_layers: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_ksp_sliders: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_ksp_slides: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_links: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_options: Data: 7.28MB + Index: 0.49MB + Engine MyISAM
wp_pmxe_exports: Data: 0.37MB + Index: 0.00MB + Engine MyISAM
wp_pmxe_google_cats: Data: 0.26MB + Index: 0.05MB + Engine MyISAM
wp_pmxe_posts: Data: 0.25MB + Index: 0.11MB + Engine MyISAM
wp_pmxe_templates: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_pmxi_files: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_pmxi_hash: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_pmxi_history: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_pmxi_images: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_pmxi_imports: Data: 0.07MB + Index: 0.00MB + Engine MyISAM
wp_pmxi_posts: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_pmxi_templates: Data: 0.02MB + Index: 0.00MB + Engine MyISAM
wp_postmeta: Data: 27.52MB + Index: 11.22MB + Engine MyISAM
wp_posts: Data: 16.15MB + Index: 1.21MB + Engine MyISAM
wp_smush_dir_images: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_term_relationships: Data: 0.01MB + Index: 0.03MB + Engine MyISAM
wp_term_taxonomy: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_termmeta: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_terms: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_tm_taskmeta: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_tm_tasks: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_usermeta: Data: 0.27MB + Index: 0.15MB + Engine MyISAM
wp_users: Data: 0.01MB + Index: 0.02MB + Engine MyISAM
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.01MB + Engine MyISAM
wp_wc_admin_notes: Data: 0.09MB + Index: 0.00MB + Engine MyISAM
wp_wc_category_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_customer_lookup: Data: 0.11MB + Index: 0.08MB + Engine MyISAM
wp_wc_download_log: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_order_addresses: Data: 0.66MB + Index: 0.76MB + Engine MyISAM
wp_wc_order_coupon_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_order_operational_data: Data: 0.73MB + Index: 0.25MB + Engine MyISAM
wp_wc_order_product_lookup: Data: 0.46MB + Index: 0.36MB + Engine MyISAM
wp_wc_order_stats: Data: 0.35MB + Index: 0.21MB + Engine MyISAM
wp_wc_order_tax_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_orders: Data: 2.52MB + Index: 1.14MB + Engine InnoDB
wp_wc_orders_meta: Data: 15.55MB + Index: 16.11MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_product_download_directories: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_product_meta_lookup: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_rate_limits: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_reserved_stock: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wc_tax_rate_classes: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wc_webhooks: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wcpdf_invoice_number: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wcpdf_packing_slip_number: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wpfb_post_templates: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_wpfb_reviews: Data: 0.10MB + Index: 0.01MB + Engine MyISAM
wp_wpfb_total_averages: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_wpml_mails: Data: 26.52MB + Index: 0.06MB + Engine InnoDB
wp_wpmm_subscribers: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_yoast_indexable: Data: 0.04MB + Index: 0.02MB + Engine MyISAM
wp_yoast_indexable_hierarchy: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_yoast_migrations: Data: 0.00MB + Index: 0.01MB + Engine MyISAM
wp_yoast_primary_term: Data: 0.00MB + Index: 0.00MB + Engine MyISAM
wp_yoast_seo_links: Data: 0.04MB + Index: 0.01MB + Engine MyISAM Post Type Counts

aam_policy: 1
attachment: 173
aw_workflow: 3
custom_css: 1
flamingo_contact: 528
flamingo_inbound: 449
hurrytimer_countdown: 1
iwp_log: 4266
kadence_element: 1
kadence_wootemplate: 3
nav_menu_item: 40
oembed_cache: 1
page: 18
post: 25
product: 10
product_variation: 2
revision: 799
shop_coupon: 10
shop_order: 4768
shop_order_refund: 72
wp_navigation: 1
wpcf7_contact_form: 1

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔ Active Plugins (39)

InfiniteWP - Client: by Revmakx – 1.13.0
Advanced Access Manager: by AAM – 6.9.42
Activity Log: by Activity Log Team – 2.11.0
Autoptimize: by Frank Goossens (futtta) – 3.1.12
Broken Link Checker: by WPMU DEV – 2.4.1
Checkout Countdown for WooCommerce: by Puri.io – 4.0.0
Contact Form 7: by Takayuki Miyoshi – 5.9.8
Flamingo: by Takayuki Miyoshi – 2.5
Google for WooCommerce: by WooCommerce – 2.8.6
Kadence Blocks - PRO Extension: by Kadence WP – 2.5.0
Kadence Blocks – Gutenberg Blocks for Page Builder Features: by Kadence WP – 3.3.2
Kadence Pro - Premium addon for the Kadence Theme: by Kadence WP – 1.0.21
Kadence Captcha: by Kadence WP – 1.3.2
Kadence WooCommerce Email Designer: by Kadence WP – 1.5.13
Limit Login Attempts Reloaded: by Limit Login Attempts Reloaded – 2.26.13
PW WooCommerce Copy Coupon: by Pimwick
Side Cart WooCommerce: by XootiX – 2.5.5
Hide SEO Bloat: by SO WP – 4.0.2
Woo Stickers by Webline: by Weblineindia – 1.2.3
PDF Invoices & Packing Slips for WooCommerce: by WP Overnight – 3.8.8
Subscribe to Newsletter for WooCommerce: by Kestrel – 4.1.1
WooCommerce: by Automattic – 9.3.3
Yoast SEO: by Team Yoast – 23.5
WP All Export Pro: by Soflyy – 1.9.1
WP Google Review Slider: by LJ Apps – 15.5
WP Last Login: by Konstantin Obenland – 7
WP Mail Logging: by WP Mail Logging Team – 1.13.1
WP-Optimize - Clean, Compress, Cache: by David Anderson
Ruhani Rabin
Team Updraft – 3.6.0
Smush: by WPMU DEV – 3.16.6
WP All Export - User Export Add-On Pro: by Soflyy – 1.0.7
WP All Export - WooCommerce Export Add-On Pro: by Soflyy – 1.0.9 Inactive Plugins (5)
Kadence Shop Kit: by Kadence WP – 2.2.6
LightStart - Maintenance Mode, Coming Soon and Landing Page Builder: by Themeisle – 2.6.13

Legacy API Enabled: –
Force SSL: –
Currency: NZD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
external (external)
grouped (grouped)
simple (simple)
variable (variable)
Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: ✔ Logging

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 90 days
Level threshold: Warning
Log directory size: 950 KB WC Pages

Shop base: # 326 - /shop/
Cart: # 327 - /shop/cart/ - Contains the [woocommerce_cart] shortcode
Checkout: # 328 - /shop/checkout/ - Contains the [woocommerce_checkout] shortcode
My account: # 329 - /shop/my-account/
Terms and conditions: # 600 - /shop/terms-conditions/ Theme

Name: === Kadence Child Theme ===
Version: 1.32
Author URL:
Child Theme: ✔
Parent Theme Name: Kadence
Parent Theme Version: 1.2.9
Parent Theme Author URL: https://www.kadencewp.com/
WooCommerce Support: ✔ Templates

Overrides: – Admin

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
mobile-app-banner
navigation
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store
Disabled Features: experimental-blocks
marketing
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
blueprint
reactify-classic-payments-settings

Daily Cron: ✔ Next scheduled: 2024-10-24 03:14:49 +13:00
Options: ✔
Notes: 175
Onboarding: completed Action Scheduler

Complete: 4,348
Oldest: 2024-09-23 13:56:20 +1200
Newest: 2024-10-24 11:11:50 +1300

Failed: 1
Oldest: 2024-10-05 16:12:07 +1300
Newest: 2024-10-05 16:12:07 +1300

Pending: 206
Oldest: 2024-10-24 11:56:05 +1300
Newest: 2025-03-07 23:59:59 +1300 Newsletter Subscription

Generated at: 2024-10-24 11:42:35 +13:00

**Comment by KevinPlusPlus** [comment]
@mattallan 

To answer your questions - 

1) I don't think we are doing anything unusual with transients, though for what it's worth, we are using Litespeed and have object caching enabled.

2) How do I enable Stripe logging? I see an option for Error logging in the Stripe plugins but I don't think that's what you're looking for.

When I was debugging this I did at one point put in backtrace calls in the Add Note callback, and if I remember correctly it did seem like the Add Note was being called again or in parallel. Maybe it would help if I added that and looked at the callbacks under 8.7 and we can compare that to 8.8.1?

Kevin

**Comment by KevinPlusPlus** [comment]
@mattallan 

Here are screenshots and backtraces to the Add Order Note function call on the subscription renewal order from when it works properly using 8.7 and when it duplicates the notes and emails in 8.8.1. Maybe seeing the backtraces of what is being processed will help.

I'm glad to insert backtraces into other places as well if that would be helpful, but since the Notes were the most obvious place on the front end, those are obviously coming from somewhere and hopefully this gives a clue.

Thanks,

Kevin

!![Image](https://github.com/user-attachments/assets/9c1ad16e-edd8-4942-99b2-f3ea788c0b47)
[Double - Backtrace.txt](https://github.com/user-attachments/files/17559381/Double.-.Backtrace.txt)
[Single - Backtrace.txt](https://github.com/user-attachments/files/17559382/Single.-.Backtrace.txt)
![Image](https://github.com/user-attachments/assets/199f70de-6c9f-4dac-b1f4-691d7d817f92)

**Comment by bdawson-ssapopka** [comment]
WooCommerce 9.3.3
WooCommerce Stripe Gateway 8.8.0
WooCommerce Memberships 1.26.9
WooCommerce Subscriptions 6.8.0

I was experiencing this issue with the new checkout experience enabled, but not when legacy checkout experience was enabled. We updated to WooCommerce Stripe Gateway 8.8.0 and turned on new checkout experience once the issue was marked as fixed, but we were still having the same symptoms in that version. I just switched our site back to the legacy checkout experience to stop duplicate order notes and emails, but the issue persists. Maybe there were some changes in the last update that cause the duplicate webhook processing to happen now no matter what checkout experience is turned on?

**Comment by n9yty** [comment]
@mattallan 

We don't have anything expiring transients that I know of. Like @KevinPlusPlus we are using Litespeed cache with object caching enabled if that helps, it is a common point anyway.

I had stripe logging turned on and I will pass them along to you.

**Comment by mattallan** [comment]
Thanks everyone for the latest responses!

I've been able to replicate two different cases which can result in duplicate `$order->payment_complete()` calls and duplicate order notes being added.

**1. Processing subscription renewals**

When we process subscription renewals, there's a small window of opportunity where the renewal order still has the "Pending" status and Stripe has sent the `payment_intent.succeeded` webhook.

This results in both our webhook handler and our process renewal functions from marking the order as completed.

**2. Conflict with Caching plugins**

With the LiteSpeed Cache plugin active on my test store, I immediately started to see duplicate order notes on new orders purchased with a new saved card. Deactivating the LiteSpeed Cache plugin on my store fixed this issue, but I need to spend a bit more time looking deeper into what's causing this.

![Image](https://github.com/user-attachments/assets/80a385d1-7079-45c4-99a3-e34bf5b221fb)

**Comment by solaceten** [comment]
Hello @mattallan thanks for the update - one problem (or rather two) is that I don't use subscriptions and I don't have lightspeed cache

**Comment by KevinPlusPlus** [comment]
@mattallan Awesome, sounds like we're getting to where you can reproduce it, and I know that's a majority of the battle sometimes. :) Let me know if I can test anything else specific. Thank you!!

@solaceten Obviously I don't know your setup, but there may be some caching layer you're not aware of that function similarly to Litespeed. So I wouldn't necessarily give up hope that solving the Litespeed issue wouldn't also solve your issue.

**Comment by danieleperilli** [comment]
Same problem. We have WooCommerce and WooCommerce Subscriptions and on each completed renewal order, the emails are duplicated. In our case, this is a very bad situation because we also have a plugin that generates invoices on QuickBooks Online, and since two different hooks of the Stripe plugin are called at the same time, sometimes this plugin creates two different invoices. As someone else suggested here, downgrading to 8.7.0 solves the problem.

This is a backtrace log that can help you understand that two different hooks are called at the same time. In this case I forced the renewal directly from the subscription action (Process Renewal), but it is the same when the renewal is automatic. Also note that to test this in a local environment, I had to use stripe CLI, otherwise webhooks are never called.

LOG
```
[31-Oct-2024 05:57:02 UTC] woo_qbo_attach_invoice_to_email:
#0 woo_qbo_create_invoice_file called at [/wp-plugins/woo-qbo/includes/invoices.php:2326]
#1 woo_qbo_attach_invoice_to_email called at [/wp-includes/class-wp-hook.php:326]
#2 apply_filters called at [/wp-includes/plugin.php:205]
#3 apply_filters called at [/wp-plugins/woocommerce/includes/emails/class-wc-email.php:477]
#4 get_attachments called at [/wp-plugins/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/emails/class-wcs-email-completed-renewal-order.php:100]
#5 trigger called at [/wp-includes/class-wp-hook.php:324]
#6 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#7 do_action called at [/wp-includes/plugin.php:517]
#8 do_action called at [/wp-plugins/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-email.php:153]
#9 send_renewal_order_email called at [/wp-includes/class-wp-hook.php:326]
#10 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#11 do_action called at [/wp-includes/plugin.php:517]
#12 do_action called at [/wp-plugins/woocommerce/includes/class-wc-order.php:411]
#13 status_transition called at [/wp-plugins/woocommerce/includes/class-wc-order.php:253]
#14 save called at [/wp-plugins/woocommerce/includes/class-wc-order.php:151]
#15 payment_complete called at [/wp-plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php:586]
#16 process_response called at [/wp-plugins/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php:557]
#17 process_subscription_payment called at [/wp-plugins/woocommerce-gateway-stripe/includes/compat/trait-wc-stripe-subscriptions.php:353]
#18 scheduled_subscription_payment called at [/wp-includes/class-wp-hook.php:324]
#19 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#20 do_action called at [/wp-includes/plugin.php:517]
#21 do_action called at [/wp-plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php:120]
#22 trigger_gateway_renewal_payment_hook called at [/wp-plugins/woocommerce-subscriptions/includes/gateways/class-wc-subscriptions-payment-gateways.php:94]
#23 gateway_scheduled_subscription_payment called at [/wp-includes/class-wp-hook.php:324]
#24 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#25 do_action called at [/wp-includes/plugin.php:517]
#26 do_action called at [/wp-plugins/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/admin/class-wcs-admin-meta-boxes.php:244]
#27 process_renewal_action_request called at [/wp-includes/class-wp-hook.php:324]
#28 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#29 do_action called at [/wp-includes/plugin.php:517]
#30 do_action called at [/wp-plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-order-actions.php:173]
#31 save called at [/wp-includes/class-wp-hook.php:324]
#32 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#33 do_action called at [/wp-includes/plugin.php:517]
#34 do_action called at [/wp-plugins/woocommerce/includes/admin/class-wc-admin-meta-boxes.php:263]
#35 save_meta_boxes called at [/wp-includes/class-wp-hook.php:326]
#36 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#37 do_action called at [/wp-includes/plugin.php:517]
#38 do_action called at [/wp-includes/post.php:4891]
#39 wp_insert_post called at [/wp-includes/post.php:4993]
#40 wp_update_post called at [/wp-admin/includes/post.php:453]
#41 edit_post called at [/wp-admin/post.php:227]

[31-Oct-2024 05:57:02 UTC] woo_qbo_attach_invoice_to_email:
#0 woo_qbo_create_invoice_file called at [/wp-plugins/woo-qbo/includes/invoices.php:2326]
#1 woo_qbo_attach_invoice_to_email called at [/wp-includes/class-wp-hook.php:326]
#2 apply_filters called at [/wp-includes/plugin.php:205]
#3 apply_filters called at [/wp-plugins/woocommerce/includes/emails/class-wc-email.php:477]
#4 get_attachments called at [/wp-plugins/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/emails/class-wcs-email-completed-renewal-order.php:100]
#5 trigger called at [/wp-includes/class-wp-hook.php:324]
#6 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#7 do_action called at [/wp-includes/plugin.php:517]
#8 do_action called at [/wp-plugins/woocommerce-subscriptions/vendor/woocommerce/subscriptions-core/includes/class-wc-subscriptions-email.php:153]
#9 send_renewal_order_email called at [/wp-includes/class-wp-hook.php:326]
#10 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#11 do_action called at [/wp-includes/plugin.php:517]
#12 do_action called at [/wp-plugins/woocommerce/includes/class-wc-order.php:411]
#13 status_transition called at [/wp-plugins/woocommerce/includes/class-wc-order.php:253]
#14 save called at [/wp-plugins/woocommerce/includes/class-wc-order.php:151]
#15 payment_complete called at [/wp-plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php:586]
#16 process_response called at [/wp-plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-webhook-handler.php:964]
#17 process_payment_intent_success called at [/wp-plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-webhook-handler.php:1192]
#18 process_webhook called at [/wp-plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-webhook-handler.php:107]
#19 check_for_webhook called at [/wp-includes/class-wp-hook.php:324]
#20 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#21 do_action called at [/wp-includes/plugin.php:517]
#22 do_action called at [/wp-plugins/woocommerce/src/Internal/Utilities/LegacyRestApiStub.php:162]
#23 maybe_process_wc_api_query_var called at [/wp-plugins/woocommerce/src/Internal/Utilities/LegacyRestApiStub.php:91]
#24 parse_legacy_rest_api_request called at [[internal function]:[no line]]
#25 call_user_func_array called at [/wp-plugins/woocommerce/src/Internal/Traits/AccessiblePrivateMethods.php:178]
#26 __callStatic called at [/wp-includes/class-wp-hook.php:324]
#27 apply_filters called at [/wp-includes/class-wp-hook.php:348]
#28 do_action called at [/wp-includes/plugin.php:565]
#29 do_action_ref_array called at [/wp-includes/class-wp.php:418]
#30 parse_request called at [/wp-includes/class-wp.php:813]
#31 main called at [/wp-includes/functions.php:1336]
#32 wp called at [/wp-blog-header.php:16]
#33 require called at [/index.php:17]
```

**Comment by ericfri** [comment]
8972987-zen

**Comment by magnatechnology** [comment]
@mattallan our team has been troubleshooting this issue for over a week, we just narrowed it down to the WooCommerce Stripe plugin and found this thread.  We also thought it could be cache related (Nitropack in this case) but disabling hasn't had any impact on the bug.  

We're having the same issue as everyone else ever since the update to 8.8.1 - duplicate order notifications are being sent to customers and store owner when a subscription order renews.  No other emails are being duplicated.

Thank you for all of your help here!

**Comment by passatgt** [comment]
Same issue in my side. I have an invoicing plugin which generates an invoice when a payment completes, but now i have duplicated invoices, because it is called twice at the same time :) In my plugin i have a check just before i make an API call to generate the actual invoice to check if it was already generated(check for invoice number in order meta), but that check fails, so that means payment complete is getting called twice exactly the same time. Also worth noting that on one site, the invoices are generated when the order status changes to completed, so thats duplicated too not just payment complete. 

I compared the code of 8.7 with 8.8 but i don't see anything that jumps out to me that might be the cause of the issue.

**Comment by ericfri** [comment]
8953579-zen

**Comment by solaceten** [comment]
@mattallan So guys, where are we with this?  How long before a potential fix? Is it worth downgrading back to a previous version for the time being, or does that not work anymore?

**Comment by mattallan** [comment]
Hi all, sorry for the delayed response! I'm working on the fixes for this issue and should have a couple of changes ready for the team to review today.

@solaceten I don't suspect downgrading to a previous Stripe version will fix the issue but if your own testing shows better results on a previous version, then yes, it should be okay to downgrad for the time being.

@passatgt does your site have any sort of object caching enabled? The issue you describe sounds like our transient that we use to prevent orders from being processed in parallel with Stripes incoming webhook is not working as expected. We're looking at a couple of fixes for this issue to put this to rest.

**Comment by radu88bc** [comment]
Hey @mattallan,

Looking forward to get this sorted out! Not sure if it helps, but we're experiencing same issues. Our current setup involves Woocommerce Subscriptions, Stripe and Litespeed Cache.

Not sure if all of the problems are related to what you're working on, but thought it might help:
- Subscription orders having their order changed from pending payment to completed then going back to pending payment and again to completed. Due to this, our invoicing platform issues two invoices and we manually need to cancel one
- Sometimes the orders go from Proccessing to Complete and then back to Proccessing, so we manually need to go and mark the order as complete
- Woocommerce Subscriptions on hold - a renewal order gets created, client gets charged but the renewal order does not "receive" the payment notification from Stripe, thus having the renewal order status in pending payment and having the actual subscription in "On Hold" status 

Note that all of the above is happening on and off.

Regards!

**Comment by Speckytoe** [comment]
I'm having double order notes. - " Order status changed from Pending payment to Processing. "

**Comment by mattallan** [comment]
Hi everyone, we've just released Stripe 8.8.2 which includes 3 different fixes which i'm 99% sure will put these duplicate email/order notes issues to rest. Can you please update to the latest Stripe version and let me know if you're still having these issues?

cc @KevinPlusPlus @dwh10045 @solaceten @passatgt @magnatechnology 

@radu88bc the issues you describe sound like they could be different issues, but some of them could still potentially fixed by our changes.

- The pending -> completed -> pending -> completed issue should be fixed on the latest Stripe (8.8.2).
- The orders that are sometimes going processing -> completed -> processing, could be related to this issue as well. Do you have some extension that is automatically marking orders as completed? Object caching could be playing a role here and an older version of an order is being saved back to the database, but it's hard to track where this could be happening. I would suggest upgrading to our Stripe 8.8.2 and see if you're still getting the issue.
- Subscriptions remaining on-hold. This one seems like a different issue. Can you please share a screenshot of the Subscriptions order notes as well as the order notes on the renewal order?

**Comment by KevinPlusPlus** [comment]
@mattallan 

Good news, the 8.8.2 update seems to have taken care of all of the duplicate notes and duplicate emails in our WooCommerce store! I guess something about the transients and object caching was just not working as expected.

Thank you very much!!!

Kevin

**Comment by solaceten** [comment]
@mattallan  Thank you for the swift work. 

 I am going to test and come back to update when I have an update.

**Comment by danieleperilli** [comment]
I do confirm that the problem has disappeared with the new version. Thanks!

**Comment by solaceten** [comment]
@mattallan - I am still getting two duplicate order emails (anybody else?).  

I'll do some more checking - but wondered if I should try to clear transients or other tasks to help resolve it ?

**Comment by mattallan** [comment]
> I am still getting two duplicate order emails

@solaceten please make sure you've cleared your caches.
Also you getting the duplicate order notes as well? If it's just duplicate emails, then it's possibly coming from some other plugin.

Let's first rule out that it's coming from our Stripe extension. Looking at your system status report, I don't see our Stripe extension in the list of active plugins so I just want confirm you are using our extension? https://wordpress.org/plugins/woocommerce-gateway-stripe/

Do you have Stripe logging enabled in **WooCommerce > Settings > Payments > Stripe**? If so, can you go to **WooCommerce > Status > Logs** and a download a copy of the **woocommerce-gateway-stripe** logs for when you had duplicate order emails? Please send me the file: `matt.allan@automattic.com`

Thanks.

---

Thanks @KevinPlusPlus and @danieleperilli for confirming the problem has been fixed on your end. I appreciate your patience and co-operation in helping resolve this. 🙌

**Comment by radu88bc** [comment]
Hey @mattallan 

First of all thanks for your assistance. It's been a week and during this time we have not had had a any duplicated order notes, so whoho! 

Please note that I am not familiar with how github works and I do apologies in advance if I should have taken other steps, but if it's ok with you I'd like to provide you with some of the additional issues and looking for your two cents if it's related to Stripe or not. 

Please see https://prnt.sc/JJWwPUBYxp9c - we often get these, but our business model involves sending out a weekly newsletter each Friday so we always get these sorted out until then. Most of the times, when I check the orders these get fixed by themselves as in they are already active and no further action needed on my end, but occasionally - like now, some of the orders require actions on my end as follows:
 
- subscription 12895 (as per screenshot) - on hold. It created a renewal order and in order notes I have this - https://prnt.sc/ySPaF6kQXy8L - so I manually have to go to Stripe and check if this customer has made a payment - which she did and I have to change the order status from pending payment to complete which also makes the subscription change from on hold to active
 
- subscription 6171 (as per screenshot) - on hold. It created a renewal order and in order notes I have this - https://prnt.sc/vBfp5sphdRLk - The payment did went through, but for some reason the order status changed from pending payment to proccessing but never got to completed. For this particular order I had to change the order status from proccessing to completed but in this particular example even though I did set the renewal order to completed, the parent subscription is still on hold and I manually have to change it to active

- subscription 6183 (as per screenshot) - on hold. This is exactly as subscription #12895 from above - after I manually checked it and changed the renewal order from pending payment to complete, the subscription also changed from on hold to active

- subscription 6309 (as per screenshot) - on hold. Again, exactly as #12895 from above with same manual fix

- subscription 6482 (as per screenshot) - on hold. Also the same as #12895 from above with same manual fix

Another example (which is not from the screenshot) is subscription 13509 - this subscription created a renewal order but there are no order notes - see https://prnt.sc/fAbesHYSw-DS and in Stripe's backend this client has already been charged for this order, so the fix here is to manually change the renewal order from pending payment to complete which changes the subscription from on hold to active.

Lastly example is subscription 9116 which created a renewal order, there are no order notes but in this particular example there has not been a charge on the customer either compared to #13509 from above, so the fix here is to manually attempt to charge the client

All the other subscriptions from the initial screenshot are are active and working as expected

Again, I do apologise if it's not the place or the method for my questions.

Regards,
Radu

**Comment by solaceten** [comment]
Hey @mattallan thank you for your help here.  I finally had some time to do some more digging and I found the issue is not caused by Stripe - it is the WooCommerce PDF Product Vouchers plugin!  When I disable it, the double email issue is gone.

**Comment by mattallan** [comment]
Hi @radu88bc so sorry for missing your message.
> https://prnt.sc/JJWwPUBYxp9c 

This admin notice indicates that there are fatal errors being thrown while processing subscription renewals. This would explain the other symptoms you've shared in your message like orders being paid on Stripe's end but not being marked as processing in Woo.

Do have PHP debug logs available?
It's also possible that your Stripe requests are timing out which could also lead to similar issues.

I realise this has been a month since your message, but if this is still happening, please go to **WooCommerce > Status > Scheduled Actions**, filter the table by failed statuses and search for the hook named "woocommerce_scheduled_subscription_payment".

Wihin the log column, this should explain why our scheduled actions are failing when processing a subscription renewal.

Happy to continue this via email: `matt.allan@automattic.com`

---

Thanks everyone for your patience while we fixed this issue. We haven't had any new reports of this so I'm going to close this issue off now 🎉
